{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "54822983",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:100% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "be7a387e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import AlignIO\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from collections import Counter\n",
    "import matplotlib.pyplot as plt\n",
    "import pickle\n",
    "from scipy.stats import fisher_exact\n",
    "from statsmodels.stats import multitest"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfd61666",
   "metadata": {},
   "source": [
    "In this analysis I'm testing the significance of SNP enrichment within clusters including all SNPs no matter how many times they occurred independently in the tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c3a082e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in summary of nonsynonymous hits in dominant clusters:\n",
    "homoplasy_dominant_clust_enrich=pickle.load(open( \"../vars/homoplasy_dominant_cluster_enrichment\", \"rb\" ) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "973edb9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "homoplasy_dominant_clust_enrich=homoplasy_dominant_clust_enrich.sort_values(by='enrichment_p', ascending=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "691bb0ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Position</th>\n",
       "      <th>major</th>\n",
       "      <th>minor</th>\n",
       "      <th>Gene</th>\n",
       "      <th>homoplasy_count</th>\n",
       "      <th>Type</th>\n",
       "      <th>Product</th>\n",
       "      <th>enrichment_p</th>\n",
       "      <th>OR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>53445</td>\n",
       "      <td>4770807</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>MAB_4690c</td>\n",
       "      <td>2</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>Linear gramicidin synthase subunit D</td>\n",
       "      <td>1.923486e-71</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>47313</td>\n",
       "      <td>4211627</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>MAB_4147c</td>\n",
       "      <td>2</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>3.635388e-69</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13841</td>\n",
       "      <td>1217813</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>MAB_1203</td>\n",
       "      <td>3</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>Mycothiol S-conjugate amidase</td>\n",
       "      <td>4.254548e-67</td>\n",
       "      <td>2914.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>38491</td>\n",
       "      <td>3381893</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>MAB_3334c</td>\n",
       "      <td>4</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>Aspartyl/glutamyl-tRNA(Asn/Gln) amidotransferase subunit B</td>\n",
       "      <td>1.034483e-65</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>24054</td>\n",
       "      <td>2119889</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>MAB_2120c</td>\n",
       "      <td>6</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>Putative inactive phenolphthiocerol synthesis polyketide synthase type I Pks15</td>\n",
       "      <td>1.709089e-64</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54903</th>\n",
       "      <td>12610</td>\n",
       "      <td>1098819</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>MAB_1086</td>\n",
       "      <td>1</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>UTP--glucose-1-phosphate uridylyltransferase</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54904</th>\n",
       "      <td>45848</td>\n",
       "      <td>4082630</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>MAB_4037</td>\n",
       "      <td>2</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54905</th>\n",
       "      <td>45849</td>\n",
       "      <td>4082799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>MAB_4037</td>\n",
       "      <td>2</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54906</th>\n",
       "      <td>45846</td>\n",
       "      <td>4082354</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>MAB_4037</td>\n",
       "      <td>2</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54907</th>\n",
       "      <td>16604</td>\n",
       "      <td>1411962</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>MAB_1408c</td>\n",
       "      <td>4</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>Putative short-chain type dehydrogenase/reductase/MSMEI_5872</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.050505</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54908 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       index Position major minor       Gene  homoplasy_count        Type  \\\n",
       "0      53445  4770807     G     A  MAB_4690c                2  intragenic   \n",
       "1      47313  4211627     C     T  MAB_4147c                2  intragenic   \n",
       "2      13841  1217813     C     T   MAB_1203                3  intragenic   \n",
       "3      38491  3381893     C     G  MAB_3334c                4  intragenic   \n",
       "4      24054  2119889     G     T  MAB_2120c                6  intragenic   \n",
       "...      ...      ...   ...   ...        ...              ...         ...   \n",
       "54903  12610  1098819     C     T   MAB_1086                1  intragenic   \n",
       "54904  45848  4082630     A     T   MAB_4037                2  intragenic   \n",
       "54905  45849  4082799     A     G   MAB_4037                2  intragenic   \n",
       "54906  45846  4082354     G     A   MAB_4037                2  intragenic   \n",
       "54907  16604  1411962     T     G  MAB_1408c                4  intragenic   \n",
       "\n",
       "                                                                              Product  \\\n",
       "0                                                Linear gramicidin synthase subunit D   \n",
       "1                                                                hypothetical protein   \n",
       "2                                                       Mycothiol S-conjugate amidase   \n",
       "3                          Aspartyl/glutamyl-tRNA(Asn/Gln) amidotransferase subunit B   \n",
       "4      Putative inactive phenolphthiocerol synthesis polyketide synthase type I Pks15   \n",
       "...                                                                               ...   \n",
       "54903                                    UTP--glucose-1-phosphate uridylyltransferase   \n",
       "54904                                                            hypothetical protein   \n",
       "54905                                                            hypothetical protein   \n",
       "54906                                                            hypothetical protein   \n",
       "54907                    Putative short-chain type dehydrogenase/reductase/MSMEI_5872   \n",
       "\n",
       "       enrichment_p           OR  \n",
       "0      1.923486e-71          inf  \n",
       "1      3.635388e-69          inf  \n",
       "2      4.254548e-67  2914.000000  \n",
       "3      1.034483e-65          inf  \n",
       "4      1.709089e-64          inf  \n",
       "...             ...          ...  \n",
       "54903  1.000000e+00          NaN  \n",
       "54904  1.000000e+00          NaN  \n",
       "54905  1.000000e+00          NaN  \n",
       "54906  1.000000e+00          NaN  \n",
       "54907  1.000000e+00     1.050505  \n",
       "\n",
       "[54908 rows x 10 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.set_option('display.max_colwidth', None)\n",
    "homoplasy_dominant_clust_enrich.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a9b37734",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/nac18/anaconda3/envs/homoplasy/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3072: DtypeWarning: Columns (7,8,9) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "# first let's merge with the mutation events table to get the change information so I can confirm which changes are synonymous\n",
    "recombFree_mutation_events=pd.read_csv('../snppar_output/mab_recombinationFree/mab_recombFree_snppar_all_mutation_events.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8329ecdd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Type</th>\n",
       "      <th>Ancestor_Node</th>\n",
       "      <th>Derived_Node</th>\n",
       "      <th>Ancestor_Call</th>\n",
       "      <th>Derived_Call</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Codon</th>\n",
       "      <th>Codon_Position</th>\n",
       "      <th>...</th>\n",
       "      <th>Derived_Codon</th>\n",
       "      <th>Ancestor_A.A.</th>\n",
       "      <th>Derived_A.A.</th>\n",
       "      <th>Change</th>\n",
       "      <th>Up_Gene</th>\n",
       "      <th>Up_Gene_Strand</th>\n",
       "      <th>Up_Gene_Distance</th>\n",
       "      <th>Down_Gene</th>\n",
       "      <th>Down_Gene_Strand</th>\n",
       "      <th>Down_Gene_Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>87</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>NODE_0000002</td>\n",
       "      <td>SAMEA2259646</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>MAB_0001</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>TCA</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>87</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>NODE_0000013</td>\n",
       "      <td>SAMEA2071363</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>MAB_0001</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>TCA</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>87</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>NODE_0000039</td>\n",
       "      <td>NODE_0000040</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>MAB_0001</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>TCA</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>87</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>NODE_0000057</td>\n",
       "      <td>NODE_0000058</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>MAB_0001</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>TCA</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>87</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>NODE_0000061</td>\n",
       "      <td>SAMN09758982</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>MAB_0001</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>TCA</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Position        Type Ancestor_Node  Derived_Node Ancestor_Call  \\\n",
       "0        87  intragenic  NODE_0000002  SAMEA2259646             G   \n",
       "1        87  intragenic  NODE_0000013  SAMEA2071363             G   \n",
       "2        87  intragenic  NODE_0000039  NODE_0000040             G   \n",
       "3        87  intragenic  NODE_0000057  NODE_0000058             G   \n",
       "4        87  intragenic  NODE_0000061  SAMN09758982             G   \n",
       "\n",
       "  Derived_Call      Gene Strand Codon Codon_Position  ... Derived_Codon  \\\n",
       "0            A  MAB_0001      1    29              3  ...           TCA   \n",
       "1            A  MAB_0001      1    29              3  ...           TCA   \n",
       "2            A  MAB_0001      1    29              3  ...           TCA   \n",
       "3            A  MAB_0001      1    29              3  ...           TCA   \n",
       "4            A  MAB_0001      1    29              3  ...           TCA   \n",
       "\n",
       "  Ancestor_A.A. Derived_A.A. Change Up_Gene Up_Gene_Strand Up_Gene_Distance  \\\n",
       "0             S            S      S       -              -                -   \n",
       "1             S            S      S       -              -                -   \n",
       "2             S            S      S       -              -                -   \n",
       "3             S            S      S       -              -                -   \n",
       "4             S            S      S       -              -                -   \n",
       "\n",
       "  Down_Gene Down_Gene_Strand Down_Gene_Distance  \n",
       "0         -                -                  -  \n",
       "1         -                -                  -  \n",
       "2         -                -                  -  \n",
       "3         -                -                  -  \n",
       "4         -                -                  -  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recombFree_mutation_events.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e07af874",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>major</th>\n",
       "      <th>minor</th>\n",
       "      <th>Position</th>\n",
       "      <th>Change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>87</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>162</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>192</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>201</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>219</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   major minor  Position Change\n",
       "0      G     A        87      S\n",
       "6      T     C       162      S\n",
       "10     G     C       192      S\n",
       "12     G     A       201      S\n",
       "18     G     C       219      S"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_events_sparse=recombFree_mutation_events[['Ancestor_Call', 'Derived_Call', 'Position', 'Change']].rename(columns={\"Ancestor_Call\": \"major\", \"Derived_Call\": \"minor\"})\n",
    "mut_events_sparse=mut_events_sparse.drop_duplicates()\n",
    "mut_events_sparse.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a8f5cf79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>major</th>\n",
       "      <th>minor</th>\n",
       "      <th>Gene</th>\n",
       "      <th>homoplasy_count</th>\n",
       "      <th>Type</th>\n",
       "      <th>Product</th>\n",
       "      <th>enrichment_p</th>\n",
       "      <th>OR</th>\n",
       "      <th>Change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4770807</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>MAB_4690c</td>\n",
       "      <td>2</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>Linear gramicidin synthase subunit D</td>\n",
       "      <td>1.923486e-71</td>\n",
       "      <td>inf</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4211627</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>MAB_4147c</td>\n",
       "      <td>2</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>3.635388e-69</td>\n",
       "      <td>inf</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1217813</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>MAB_1203</td>\n",
       "      <td>3</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>Mycothiol S-conjugate amidase</td>\n",
       "      <td>4.254548e-67</td>\n",
       "      <td>2914.0</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3381893</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>MAB_3334c</td>\n",
       "      <td>4</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>Aspartyl/glutamyl-tRNA(Asn/Gln) amidotransferase subunit B</td>\n",
       "      <td>1.034483e-65</td>\n",
       "      <td>inf</td>\n",
       "      <td>NS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2119889</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>MAB_2120c</td>\n",
       "      <td>6</td>\n",
       "      <td>intragenic</td>\n",
       "      <td>Putative inactive phenolphthiocerol synthesis polyketide synthase type I Pks15</td>\n",
       "      <td>1.709089e-64</td>\n",
       "      <td>inf</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Position major minor       Gene  homoplasy_count        Type  \\\n",
       "0   4770807     G     A  MAB_4690c                2  intragenic   \n",
       "1   4211627     C     T  MAB_4147c                2  intragenic   \n",
       "2   1217813     C     T   MAB_1203                3  intragenic   \n",
       "3   3381893     C     G  MAB_3334c                4  intragenic   \n",
       "4   2119889     G     T  MAB_2120c                6  intragenic   \n",
       "\n",
       "                                                                          Product  \\\n",
       "0                                            Linear gramicidin synthase subunit D   \n",
       "1                                                            hypothetical protein   \n",
       "2                                                   Mycothiol S-conjugate amidase   \n",
       "3                      Aspartyl/glutamyl-tRNA(Asn/Gln) amidotransferase subunit B   \n",
       "4  Putative inactive phenolphthiocerol synthesis polyketide synthase type I Pks15   \n",
       "\n",
       "   enrichment_p      OR Change  \n",
       "0  1.923486e-71     inf      S  \n",
       "1  3.635388e-69     inf      S  \n",
       "2  4.254548e-67  2914.0      S  \n",
       "3  1.034483e-65     inf     NS  \n",
       "4  1.709089e-64     inf      S  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "homoplasy_dominant_clust_enrich=homoplasy_dominant_clust_enrich.astype({'Position': 'int64'})\n",
    "enriched_vars_change=homoplasy_dominant_clust_enrich.merge(mut_events_sparse, how=\"left\", on=['major', 'minor', 'Position'])\n",
    "enriched_vars_change.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "bc7b3a84",
   "metadata": {},
   "outputs": [],
   "source": [
    "# divide the data up into synonymous and nonsynonymous variation\n",
    "ns_vars=enriched_vars_change[enriched_vars_change.Change=='NS']\n",
    "s_vars=enriched_vars_change[enriched_vars_change.Change=='S']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3ead8b42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.5845002079578538"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculate FDR for nonsynonymous SNPs\n",
    "ns_fdr=multitest.fdrcorrection(ns_vars.enrichment_p)\n",
    "sum(ns_fdr[0])/len(ns_fdr[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "339c1fab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.6475437605872388"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculate FDR for synonymous SNPs\n",
    "s_fdr=multitest.fdrcorrection(s_vars.enrichment_p)\n",
    "sum(s_fdr[0])/len(s_fdr[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "205fecc4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.6274623827869449"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculate FDR for all SNPs combined\n",
    "fdr=multitest.fdrcorrection(enriched_vars_change.enrichment_p)\n",
    "sum(fdr[0])/len(fdr[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "2d200f75",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1     24535\n",
       "2      9838\n",
       "3      6341\n",
       "4      4571\n",
       "5      3525\n",
       "6      2472\n",
       "7      1592\n",
       "0       841\n",
       "8       798\n",
       "9       359\n",
       "10      112\n",
       "11       38\n",
       "12        4\n",
       "20        1\n",
       "13        1\n",
       "Name: homoplasy_count, dtype: int64"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enriched_vars_change.homoplasy_count.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da49fa42",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:homoplasy]",
   "language": "python",
   "name": "conda-env-homoplasy-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
